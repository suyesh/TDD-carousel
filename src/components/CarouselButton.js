import React from 'react';

const CarouselButton = props => <button type="button" {...props} />;

export default CarouselButton;
