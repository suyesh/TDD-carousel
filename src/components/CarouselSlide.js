import React from 'react';
import PropTypes from 'prop-types';

const CarouselSlide = ({
  imgUrl,
  description,
  attribution,
  style,
  onClick,
  className
}) => {
  const handleKeyDown = () => {
    onClick();
  };

  return (
    <figure
      style={style}
      onClick={onClick}
      onKeyDown={handleKeyDown}
      className={className}
      role="presentation"
    >
      <img src={imgUrl} alt={description} />
      <figcaption>
        <strong>{description}</strong> {attribution}
      </figcaption>
    </figure>
  );
};

CarouselSlide.propTypes = {
  imgUrl: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,
  attribution: PropTypes.node,
  style: PropTypes.objectOf(PropTypes.object),
  className: PropTypes.string,
  onClick: PropTypes.func
};

export default CarouselSlide;
