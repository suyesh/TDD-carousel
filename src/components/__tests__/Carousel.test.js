import React from 'react';
import { shallow } from 'enzyme';
import Carousel from '../Carousel';
import CarouselButton from '../CarouselButton';
import CarouselSlide from '../CarouselSlide';

describe('Carousel', () => {
  let wrapper;

  const slides = [
    {
      imgUrl: 'https://example.com/slide1.png',
      description: 'Slide 1',
      attribution: 'Uno Pizzeria'
    },
    {
      imgUrl: 'https://example.com/slide2.png',
      description: 'Slide 2',
      attribution: 'Dos Equis'
    },
    {
      imgUrl: 'https://example.com/slide3.png',
      description: 'Slide 3',
      attribution: 'Three Amigos'
    }
  ];

  beforeEach(() => {
    wrapper = shallow(<Carousel slides={slides} />);
  });

  afterAll(() => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a <div></div>', () => {
    expect(wrapper.type()).toBe('div');
  });

  it('has an initial `slideIndex` of 0', () => {
    expect(wrapper.state('slideIndex')).toBe(0);
  });

  it('renders a CarouselButton labeled `prev`', () => {
    expect(
      wrapper
        .find(CarouselButton)
        .at(0)
        .prop('children')
    ).toBe('Prev');
  });

  it('renders a CarouselButton labeled `Next`', () => {
    expect(
      wrapper
        .find(CarouselButton)
        .at(1)
        .prop('children')
    ).toBe('Next');
  });

  it('decrements `slideIndex` when Prev is clicked', () => {
    wrapper.setState({ slideIndex: 1 });
    wrapper.find('[data-action="prev"]').simulate('click');
    expect(wrapper.state('slideIndex')).toBe(0);
  });

  it('increments `slideIndex` when Next is clicked', () => {
    wrapper.setState({ slideIndex: 1 });
    wrapper.find('[data-action="next"]').simulate('click');
    expect(wrapper.state('slideIndex')).toBe(2);
  });

  it('renders the current slide as a CarouselSlide', () => {
    let slideProps;
    slideProps = wrapper.find(CarouselSlide).props();
    expect(slideProps).toEqual(slides[0]);
    wrapper.setState({ slideIndex: 1 });
    slideProps = wrapper.find(CarouselSlide).props();
    expect(slideProps).toEqual(slides[1]);
  });
});
