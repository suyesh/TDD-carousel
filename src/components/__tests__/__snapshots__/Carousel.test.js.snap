// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel increments \`slideIndex\` when Next is clicked 1`] = `
<div>
  <CarouselButton
    data-action="prev"
    onClick={[Function]}
  >
    Prev
  </CarouselButton>
  <CarouselButton
    data-action="next"
    onClick={[Function]}
  >
    Next
  </CarouselButton>
</div>
`;

exports[`Carousel renders the current slide as a CarouselSlide 1`] = `
<div>
  <CarouselSlide
    attribution="Dos Equis"
    description="Slide 2"
    imgUrl="https://example.com/slide2.png"
  />
  <CarouselButton
    data-action="prev"
    onClick={[Function]}
  >
    Prev
  </CarouselButton>
  <CarouselButton
    data-action="next"
    onClick={[Function]}
  >
    Next
  </CarouselButton>
</div>
`;
